### DISCUSSION TITLE
poetry install hang

### DISCUSSION THREAD

**Discussion by nothingtodowu** [pybricks/support#1591] [General]
Hi pybricks team,

Thanks for all the efforts on pybricks! It is truly great!
I am new to Pybricks and this is the first time I try to compile code.
When I follow this, https://github.com/pybricks/pybricks-micropython/blob/master/CONTRIBUTING.md
My setup just hang there after I run "poetry install"
Anything I forgot for this issue? Thanks a lot.
Here is the log:

max@max-Default-string:~/pybricks-micropython$ poetry install
Installing dependencies from lock file

Package operations: 151 installs, 1 update, 0 removals

  - Installing jsonschema-specifications (2023.12.1): Pending...
  - Installing platformdirs (4.2.0): Pending...
  - Installing python-dateutil (2.9.0.post0): Pending...
  - Installing traitlets (5.14.1): Pending...
  - Installing types-python-dateutil (2.8.19.20240106): Pending...

Best regards,
wu

**Comment by dlech** [comment]
Sometimes poetry can take a really long time to resolve package dependencies. What Python version are you using? How long did you actually wait?

**Reply by nothingtodowu** [reply]
Thank you so much. I will try to wait long time and check.

**Reply by nothingtodowu** [reply]
Hi dlech, I had wait for 5 hours, it still keep that log, no more information. My python is 3.10.14. Is this correct?

**Reply by dlech** [reply]
Python 3.10 should be fine.

And I should have noticed this before, but if it is getting to the install part, then the dependency resolution is done. 

This (https://python-poetry.org/docs/repositories#debugging-issues) suggests running `poetry` with the `--no-cache` and `-vvv` options for troubleshooting.

**Reply by nothingtodowu** [reply]
Hi dlech,

I found the issue. My server is Ubuntu 20.04, which default python3 is python 3.8.4. Even I install python 3.10.14, python3 still link to 3.8.4. Then poetry just hang there. Can I use python 3.84 for build? Or any idea to change python3 to python3.10.14? I try several ways but all failed to install potery at my ubuntu 20.04.

**Reply by dlech** [reply]
You can type `poetry env use python3.10` to use the different Python version, then continue with `poetry install`.

**Reply by nothingtodowu** [reply]
Thanks a lot. After I re-install Ubuntu 22.04, then everything is ok. I just got another issue when I run "make primehub". Below is the error message. Where can I find this "arm-none-eabi-gcc"? Thanska lot.

Submodule path 'lib/STM32_USB_Device_Library': checked out '7b5e6886d2f11ad15d2b46364e8a96984358f639'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
mkdir -p build/genhdr
GEN build/genhdr/mpversion.h
GEN build/genhdr/qstr.i.last
Traceback (most recent call last):
  File "/home/max/pybricks-micropython/bricks/primehub/../../micropython/py/makeqstrdefs.py", line 209, in <module>
    preprocess()
  File "/home/max/pybricks-micropython/bricks/primehub/../../micropython/py/makeqstrdefs.py", line 76, in preprocess
    for output in p.imap(pp(flags), chunks):
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/max/pybricks-micropython/bricks/primehub/../../micropython/py/makeqstrdefs.py", line 60, in run
    return subprocess.check_output(args.pp + flags + files)
  File "/usr/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1863, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'arm-none-eabi-gcc'
make[1]: *** [../../micropython/py/mkrules.mk:111: build/genhdr/qstr.i.last] Error 1
make[1]: *** Deleting file 'build/genhdr/qstr.i.last'
make[1]: Leaving directory '/home/max/pybricks-micropython/bricks/primehub'
make: *** [Makefile:98: primehub] Error 2

**Reply by nothingtodowu** [reply]
Hi Dlech,
I fixed issue with this link. Now I can build primehub image. Thanks a lot.
https://stackoverflow.com/questions/73976876/e-unable-to-locate-package-arm-none-eabi-gcc

**Comment by nothingtodowu** [comment]
Hi Dlech, Now I can build image for primehub and download image now. May I know how to get this branch to latest master and build image? I try with "git branch -a | grep bluetooth" but can't find this branch. Thanks a lot.

https://github.com/nkarstens/pybricks-micropython/tree/bluetooth-toggle

**Reply by dlech** [reply]
```sh
git checkout -b nkarstens-bluetooth-toggle master
git pull https://github.com/nkarstens/pybricks-micropython.git bluetooth-toggle
```

**Reply by nothingtodowu** [reply]
Hi Dlech, 

Thank you so much. It works.
I have another question: I saw there is a branch 'dlech-usb'. I had compiled it and download it to primehub. But I can't find how to use this image for download code with USB cable. Is this branch works for USB download code? If yes, how to do that? Thanks a lot.

**Reply by dlech** [reply]
We don't have a fully working USB branch yet. https://github.com/pybricks/pybricks-micropython/pull/208 is the latest and greatest and includes changes from my branch. There is a corresponding pull request for pybricksdev to make use of the firmware changes at https://github.com/pybricks/pybricksdev/pull/69.

**Reply by nothingtodowu** [reply]
Hi Lechner,

Thanks a lot. After clone this branch to latest master, I got USB device "Lego Technic Large Hub + Pybricks". Any idea to download code with this device? I can try it. 
https://github.com/nkarstens/pybricks-micropython/tree/usb

**Reply by dlech** [reply]
poetry run pybricksdev run usb my_script.py

**Reply by nothingtodowu** [reply]
Hi Lechner,

Thanks a lot. Is there anyway for Windows to program code? Or must use Ubuntu?

**Reply by dlech** [reply]
Compiling the Pybricks firmware probably only works on Linux currently, but everything else should be working cross-platform.

**Reply by nothingtodowu** [reply]
I try with this command "poetry run pybricksdev run usb run_wro.py" at Ubuntu 22.04 with spike prime hub which download usb enable firmware. Here is the error messages I got. Why it show "Could not find hub"? I already connect hub with the spike prime hub. 

(pybricks-micropython-py3.10) nk@nk-ThinkPad-X250:~/pybricks-micropython$ poetry run pybricksdev run usb run_wro.py 
Traceback (most recent call last):
  File "/home/nk/pybricks-micropython/.venv/bin/pybricksdev", line 8, in <module>
    sys.exit(main())
  File "/home/nk/pybricks-micropython/.venv/lib/python3.10/site-packages/pybricksdev/cli/__init__.py", line 383, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/nk/pybricks-micropython/.venv/lib/python3.10/site-packages/pybricksdev/cli/__init__.py", line 204, in run
    await hub.connect(device_or_address)
  File "/home/nk/pybricks-micropython/.venv/lib/python3.10/site-packages/pybricksdev/connections/lego.py", line 86, in connect
    raise OSError("Could not find hub.")
OSError: Could not find hub.

**Reply by laurensvalk** [reply]
USB is not yet supported. The command `poetry run pybricksdev run usb my_script.py` is an outdated command that has nothing to do with this unfortunately.

> `run_wro.py` 

I would not recommend attempting to use USB for competitions at this stage. It is just not ready for that.

So backing up a few steps... Is USB absolutely needed for your case? Bluetooth is allowed in WRO as far as I know, at least according to the global organizers when I recently asked. It might be worth checking the rules of your local competitions.

**Reply by nothingtodowu** [reply]
Thanks a lot. Then I use bluetooth for now. Hope your USB official support can be released earlier.

**Comment by nothingtodowu** [comment]
Hi Lechner,

Sometime when I compile code I got this errors. How to recover when this issue happen? Thanks a lot.

```console
(pybricks-micropython-py3.10) nk@nk-ThinkPad-X250:~/pybricks-micropython$ make primehub
make[1]: Entering directory '/home/nk/pybricks-micropython/micropython/mpy-cross'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
make[1]: Leaving directory '/home/nk/pybricks-micropython/micropython/mpy-cross'
make[1]: Entering directory '/home/nk/pybricks-micropython/bricks/primehub'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Including _hub_extra as a module.
make[1]: Leaving directory '/home/nk/pybricks-micropython/bricks/primehub'
```

**Reply by nothingtodowu** [reply]
This issue happen when I install and run VS Code at Ubuntu.

**Reply by dlech** [reply]
I don't see any error. It looks like a successful build to me. No output files were updated probably because they were built already and no source files were modified.

**Reply by nothingtodowu** [reply]
Got it

**Comment by nothingtodowu** [comment]
Hi Lechner,

Sorry, another issue. When I download firmware.zip at Ubuntu, chrome, I got this error. What need I do for Ubuntu download firmware? Thanks a lot.

An unexpected error occurred. Please consider reporting this so we can make a better error message.

Failed to execute 'open' on 'USBDevice': Access denied.

Expand for detailed technical information
Error: Failed to execute 'open' on 'USBDevice': Access denied.

**Reply by dlech** [reply]
Do you have the proper udev rules installed?

**Reply by nothingtodowu** [reply]
How to install this udev rules? My setup is Ubuntu 22.04

**Reply by dlech** [reply]
There is a link in Pybricks Code to the instructions.

![image](https://github.com/pybricks/support/assets/963645/c0cea2b8-394e-4bf4-bdb1-29dbc000983b)

https://pybricks.com/projects/tutorials/dev/tools/linux/

**Reply by nothingtodowu** [reply]
Thanks a lot
